text/0040-pagination-convention.md

```diff
@@ -0,0 +1,118 @@
+- Title: Pagination Convention
+- Start Date: 2021-05-10
+- Specification PR: [#40](https://github.com/meilisearch/specifications/pull/40)
+- MeiliSearch Tracking-Issues:
+
+# Pagination Convention
+
+## 1. Feature Description and Interaction
+
+### I. Summary
+
+Pagination allows the content to be viewed in the form of a sequence in order to facilitate the browsing of its content by not overloading the number of elements to be viewed for the user. Pagination also has the advantage of easily providing the total number of items that remain to be traversed by not overloading the network with all the content that can be loaded.
+
+### II. Motivation
+
+This specification will be used as a convention to implement simple and standard paginations to be set up on API routes such as the list of updates for example. Since the pagination of search results is intrinsically linked to technical implementations of the search engine, this specific case is detailed in its own specification.
+
+### III. Additional Materials
+N/A
+
+### IV. Explanation
+
+##### Requirements
+
+- ✅ Each paginable route must return the total number of items it can render to the user at the end and the current number of items returned in the response. E.g. `total` and `size`. In the case of multiples pages, all other pages except the last one will have a `size = limit`.
+- ✅ Each paginable route should be able to handle `offset` and `limit` to return a specific page.
+- ✅ Each paginable route should handle a default `limit`. E.g. `20`
+- ✅ Each paginable route should handle a default `offset` parameter with a value of `0`.
+- ✅ `limit` and `offset` are number parameters.
+- ✅ If `limit` and `offset` can't return any results, the API should not return an error. It's preferable to return empty results in this case.
+
+##### Going further
+
+- Provide a `_links` object with `current`, `next` and `prev` elements to facilitate pagination on client side by pre-computing `next` and `prev` urls parameters given the pagination context.
```

Comment:
Do you think it really helps the user? I have the impression that it is not used too much by API consumers. If I take the example of the [Stripe API](https://stripe.com/docs/api/pagination), the only information they add in the response for paging is the `has_more`.

Reply:
Thinking about it, the advantage is that it could be easier to build a pagination on "simple" frontends. The links being directly accessible, the pagination rendering component would only have to take and assign the prev, next on related pagination buttons. However, this does not meet all types of pagination and some frontend frameworks may not be able to use the links directly but have to reform this preconceived url from the state variables.

As we don't have any direct feedback on this subject, I will only leave this mention within the future possibilities.